### PR TITLE
Fix usage of custom EVP_CIPHER objects

### LINKED ISSUE #18970: Silent Failure in OpenSSL leads to use of NULL Cipher
Hello,

I know this is a corner case and one that has been warned about but I thought it would be good to report this anyway. 

Under 1.1.1 I was using a custom EVP so that we could implement a multithreaded AES_CTR in our code (hpnssh, a variant of openssh). This was defined as follows:

` const EVP_CIPHER *
evp_aes_ctr_mt(void)
{
        static EVP_CIPHER *aes_ctr;
        aes_ctr = EVP_CIPHER_meth_new(NID_undef, 16/*block*/, 16/*key*/);
        EVP_CIPHER_meth_set_iv_length(aes_ctr, AES_BLOCK_SIZE);
        EVP_CIPHER_meth_set_init(aes_ctr, ssh_aes_ctr_init);
        EVP_CIPHER_meth_set_cleanup(aes_ctr, ssh_aes_ctr_cleanup);
        EVP_CIPHER_meth_set_do_cipher(aes_ctr, ssh_aes_ctr);
        EVP_CIPHER_meth_set_flags(aes_ctr, EVP_CIPH_CBC_MODE
                                      | EVP_CIPH_VARIABLE_LENGTH
                                      | EVP_CIPH_ALWAYS_CALL_INIT
                                      | EVP_CIPH_CUSTOM_IV);
        return (aes_ctr);
}`

This compiles with deprecated warnings under OpenSSL 3 but it does compile. However, interesting things happen when we try to use it. I am not entirely sure but what believe what is happening is that when we try to init this type (evp_aes_ctr_mt) the OpenSSL code looks for a provider for this, can't find one, and possibly returns NULL. This doesn't cause EVP_CipherInit to fail though. Instead it returns 0 and it looks like everything is working. Except that when we move data across this cipher the result is that everything is en clear. I confirmed this via a tcpdump between a server and client both built against OpenSSL 3.0.2 and was able to observe all of the data being transmitted across the wire unencrypted. 

Additionally, a connection between a 1.1.1 and 3.0.2 did not work as each side saw the other as being corrupted. Which is what you'd expect if the above is true. 

I know that using any of the meth_ functions in OpenSSL 3 is "being deprecated in OpenSSL 3.0, and users of these APIs should know that their use can likely bypass provider selection and configuration, with unintended consequences." However, this seems like a severe unintended consequence. 

In any event, I'm not expecting a bug fix but I wanted to let you know in case you needed to know what some of these unintended consequences look like.

### PR DESCRIPTION
**Pull request by mattcaswell**
If a custom EVP_CIPHER object has been passed to EVP_CipherInit() then it
should be used in preference to a fetched cipher.

We also fix a possible NULL pointer deref in the same code for digests.

If the custom cipher passed to EVP_CipherInit() happens to use NID_undef
(which should be a discouraged practice), then in the previous
implementation this could result in the NULL cipher being fetched and
hence NULL encryption being unexpectedly used.

CVE-2022-3358

Fixes #18970

Note that this has been assessed as a low severity security issue. As per our policy, since it is low, it is being fixed in public and will be included in the next release.

We also fix a bug where the cleanup function for an EVP_CIPHER was not always being called when it should, which was highlighted by writing the test for the original issue.

### CODE REVIEW

File: crypto/evp/evp_enc.c [hunk 1]
```diff
@@ -156,11 +159,12 @@ static int evp_cipher_init_internal(EVP_CIPHER_CTX *ctx,
      * (legacy code)
      */
     if (cipher != NULL && ctx->cipher != NULL) {
+        if (ctx->cipher->cleanup && !ctx->cipher->cleanup(ctx))
```
Comment (t8m) [inline crypto/evp/evp_enc.c:162]: Nit: `ctx->cipher->cleanup != NULL`

File: test/evp_extra_test.c [hunk 2]
```diff
@@ -4329,6 +4329,88 @@ static int test_custom_md_meth(void)
...
+    EVP_CIPHER *tmp = NULL;
+    int testresult = 0;
+    int nid;
+
+    /*
+     * We are testing deprecated functions. We don't support a non-default
+     * library context in this test.
+     */
+    if (testctx != NULL)
+        return 1;
```
Comment (t8m) [inline test/evp_extra_test.c:4375]: Maybe `return TEST_skip("Non-default libctx");` ?

**Review by t8m — CHANGES_REQUESTED** [review]

**Comment by mattcaswell** [comment]
Feedback addressed.

**Comment by togtalk** [comment]
Yes nice please share

**Review by t8m — APPROVED** [review]

File: test/evp_extra_test.c [hunk 3]
```diff
@@ -4329,6 +4329,88 @@ static int test_custom_md_meth(void)
...
+    if (!TEST_true(EVP_CIPHER_meth_set_init(tmp, custom_ciph_init))
+            || !TEST_true(EVP_CIPHER_meth_set_flags(tmp, EVP_CIPH_ALWAYS_CALL_INIT))
+            || !TEST_true(EVP_CIPHER_meth_set_cleanup(tmp, custom_ciph_cleanup))
+            || !TEST_true(EVP_CIPHER_meth_set_impl_ctx_size(tmp,
+                                                            sizeof(custom_ciph_ctx))))
+        goto err;
+
+    ciphctx = EVP_CIPHER_CTX_new();
+    if (!TEST_ptr(ciphctx)
+               /*
```
Comment (paulidale) [inline test/evp_extra_test.c:4395]: Is the indentation off here?

**Review by paulidale — APPROVED** [review]

File: test/evp_extra_test.c (hunk 3 above)
Comment (levitte) [reply test/evp_extra_test.c:4395]: Yes it is

**Comment by paulidale** [comment]
Needs an indentation fix before merge.

**Comment by mattcaswell** [comment]
Pushed to master and 3.0. I fixed the indentation issue during merge. There was a trivial conflict cherry-picking to 3.0 that I also resolved during the merge.

**Pull request closed without merging**